metagraph/src/graph/alignment/aligner_aggregator.hpp

```diff
@@ -23,54 +34,69 @@ class AlignmentAggregator {
 
     size_t size() const { return path_queue_.size(); }
+
     bool empty() const { return path_queue_.empty(); }
 
+    void clear() { path_queue_.clear(); }
+
   private:
     std::string_view query_;
     std::string_view rc_query_;
     const DBGAlignerConfig &config_;
-    boost::container::priority_deque<DBGAlignment,
-                                     std::vector<DBGAlignment>,
-                                     AlignmentCompare> path_queue_;
+    PathQueue path_queue_;
+    AlignmentCompare cmp_;
 };
 
 
-template <typename NodeType, class AlignmentCompare>
-inline void AlignmentAggregator<NodeType, AlignmentCompare>
-::add_alignment(DBGAlignment&& alignment) {
+template <class AlignmentCompare>
+inline void AlignmentAggregator<AlignmentCompare>
+::add_alignment(Alignment&& alignment) {
+    if (std::find(path_queue_.begin(), path_queue_.end(), alignment) != path_queue_.end())
+        return;
+
     if (path_queue_.size() < config_.num_alternative_paths) {
         path_queue_.emplace(std::move(alignment));
-    } else if (!AlignmentCompare()(alignment, path_queue_.minimum())) {
+    } else if (!cmp_(alignment, path_queue_.minimum())) {
         path_queue_.update(path_queue_.begin(), std::move(alignment));
     }
 }
 
-template <typename NodeType, class AlignmentCompare>
-inline auto AlignmentAggregator<NodeType, AlignmentCompare>
-::get_min_path_score(const DBGAlignment &) const -> score_t {
-    return path_queue_.size() ? path_queue_.minimum().get_score()
-                              : config_.min_path_score;
+template <class AlignmentCompare>
+inline auto AlignmentAggregator<AlignmentCompare>::get_min_path_score() const -> score_t {
+    return path_queue_.size() < config_.num_alternative_paths
+        ? config_.min_path_score
+        : std::max(static_cast<score_t>(path_queue_.maximum().get_score() * config_.rel_score_cutoff),
+                   path_queue_.minimum().get_score());
 }
 
-template <typename NodeType, class AlignmentCompare>
-inline void AlignmentAggregator<NodeType, AlignmentCompare>
-::call_alignments(const std::function<void(DBGAlignment&&)> &callback) {
-    if (path_queue_.empty())
-        return;
+template <class AlignmentCompare>
+inline auto AlignmentAggregator<AlignmentCompare>::get_max_path_score() const -> score_t {
+    return path_queue_.size() ? path_queue_.maximum().get_score() : config_.min_path_score;
+}
 
-    while (path_queue_.size()) {
-        callback(DBGAlignment(path_queue_.maximum()));
-        path_queue_.pop_maximum();
+template <class AlignmentCompare>
+inline void AlignmentAggregator<AlignmentCompare>
+::call_alignments(const std::function<void(Alignment&&)> &callback,
+                  const std::function<bool()> &terminate) {
+    auto &data = path_queue_.data();
+    for (auto it = data.rbegin(); it != data.rend() && !terminate(); ++it) {
+        boost::heap::pop_interval_heap_max(data.begin(), it.base(), path_queue_.cmp());
+        callback(std::move(*it));
     }
 }
```

Comment:
Using iterators here would be strange since each retrieval pops an element off the list.

Reply:
Then the name `call_alignments` is confusing, it doesn't make it clear that it changes the state and can be called only once.

btw, why don't you return them in a vector?

Reply:
It could be that the termination condition is dependent on the alignments that are returned (e.g., terminate after returning three alignments, or terminate when the score is below a value X, etc.).

Reply:
but they're already computed, right? Why not always return all of them?

Reply:
The alignments are computed, but they should be returned in sorted (non-increasing score) order. So, if the queue is large, but you're only interested in, say, the top two alignments, then you only end up performing 2log(n) operations instead of nlog(n) to retrieve them.

Reply:
`call_alignments` is not used here, so I just replaced it with `get_alignments`. We can add `call_alignments` later if needed.